### PR TITLE
Implement @Type for Union

### PR DESCRIPTION
**Pull request by tadeokondrak**
I hope I didn't break unions...

### CODE REVIEW

**Comment by tadeokondrak** [comment]
Turns out some things in the standard library use TypeInfo.UnionField.enum_field and I'm running into some compiler bugs replacing them

**Review by alexnask — APPROVED** [review]

File: src/ir.cpp [hunk 1]
```diff
@@ -26390,13 +26379,84 @@ static ZigType *type_info_to_type(IrAnalyze *ira, IrInst *source_instr, ZigTypeI
                     return ira->codegen->invalid_inst_gen->value->type;
                 field->value = *field_int_value;
             }
+            return entry;
+        }
+        case ZigTypeIdUnion: {
+            assert(payload->special == ConstValSpecialStatic);
+            assert(payload->type == ir_type_info_get_type(ira, "Union", nullptr));
 
+            ZigValue *layout_value = get_const_field(ira, source_instr->source_node, payload, "layout", 0);
```
Comment (Vexu) [inline src/ir.cpp:26388]: Missing null check.

File: src/ir.cpp [hunk 2]
```diff
@@ -26390,13 +26379,84 @@ static ZigType *type_info_to_type(IrAnalyze *ira, IrInst *source_instr, ZigTypeI
...
+            assert(layout_value->special == ConstValSpecialStatic);
+            assert(layout_value->type == ir_type_info_get_type(ira, "ContainerLayout", nullptr));
+            ContainerLayout layout = (ContainerLayout)bigint_as_u32(&layout_value->data.x_enum_tag);
+
+            ZigType *tag_type = get_const_field_meta_type_optional(ira, source_instr->source_node, payload, "tag_type", 1);
```
Comment (Vexu) [inline src/ir.cpp:26393]: Missing error check: `tag_type != nullptr && type_is_invalid(tag_type)`.

File: src/ir.cpp [hunk 3]
```diff
@@ -26390,13 +26379,84 @@ static ZigType *type_info_to_type(IrAnalyze *ira, IrInst *source_instr, ZigTypeI
...
+            if (tag_type != nullptr && tag_type->id != ZigTypeIdEnum) {
+                ir_add_error(ira, source_instr, buf_sprintf(
+                    "union tag type must be an enum, not %s", type_id_name(tag_type->id)));
```
Comment (Vexu) [inline src/ir.cpp:26396]:
```suggestion
                    "expected enum type, found '%s'", type_id_name(tag_type->id)));
```

File: src/analyze.cpp [hunk 4]
```diff
@@ -3069,30 +3068,50 @@ static Error resolve_union_zero_bits(CodeGen *g, ZigType *union_type) {
...
-    assert(union_type->data.unionation.fields == nullptr);
-    uint32_t field_count = (uint32_t)decl_node->data.container_decl.fields.length;
-    if (field_count == 0) {
-        add_node_error(g, decl_node, buf_sprintf("unions must have 1 or more fields"));
+    uint32_t field_count;
+    if (decl_node->type == NodeTypeContainerDecl) {
+        assert(union_type->data.unionation.fields == nullptr);
+        field_count = (uint32_t)decl_node->data.container_decl.fields.length;
+        if (field_count == 0) {
+            add_node_error(g, decl_node, buf_sprintf("unions must have 1 or more fields"));
```
Comment (Vexu) [inline src/analyze.cpp:3076]: This should also be checked for generated unions. Seems to also be missing from enums.

File: src/analyze.cpp [hunk 5]
```diff
@@ -3150,105 +3169,111 @@ static Error resolve_union_zero_bits(CodeGen *g, ZigType *union_type) {
...
                     union_type->data.unionation.resolve_status = ResolveStatusInvalid;
                     return ErrorSemanticAnalyzeFail;
-                case ReqCompTimeYes:
-                    union_type->data.unionation.requires_comptime = true;
-                    break;
-                case ReqCompTimeNo:
-                    break;
-            }
+                }
+                if (field_is_opaque_type) {
```
Comment (Vexu) [inline src/analyze.cpp:3228]: This should also be checked for generated unions.

**Review by Vexu — CHANGES_REQUESTED** [review]

**Comment by tadeokondrak** [comment]
Rebased to fix conflict

**Pull request merged**